src/node/chainstate.cpp

```diff
@@ -57,35 +61,45 @@ static ChainstateLoadResult CompleteChainstateInitialization(
         }
     }
 
-    if (chainman.m_interrupt) return {ChainstateLoadStatus::INTERRUPTED, {}};
+    if (chainman.m_interrupt) {
+        result.Update(Interrupted{});
+        return result;
```

Comment:
> I wrote a simple patch to make the existing `Result` more flexible (allow inner error types other than `bilingual_str`), for places that want to use something like `std::expected<A, B>` in code today by using `Result<A, B>`.

I think it could be a good idea to introduce a `util::Expected<A, B>` class if you see places in our code that need to return structured failure information but do not need to return error messages along with the failure data. If we have a `util::Expected` class it should make it easier to transition code from `util::Expected` to `std::expected` (maybe even with a scripted-diff)` than if that code were written `util::Result`.

I do see benefits of `util::Result` and `std::expected` as being pretty different. I think `util::Result` class provides a good way to return complete error information for GUI, RPC, and libbitcoinkernel users when code is doing a big operation like loading a wallet, connecting a block, or creating a transaction that can fail in complicated ways. It provides a way for wallet code in #25722 and kernel code in #29700 to return error information in a standard format in a single return value without ad-hoc `bilingual_string& error` `std::vector<bilingual_str>& warnings` `bool& fatal_error` `bool& is_interrupted` in/out arguments. I think `Result` class works best when you have a chain of `Result` functions calling other `Result` functions, because the class provides ways to safely merge information for types that can be merged while triggering compile errors if there are attempts to combine incompatible results.

By contrast, `std::expected<A, B>` is basically syntax sugar over `std::variant<A, B>`, which is great, but I think doesn't help as much with problems in our codebase of returning error messages with enough context to users and making sure fatal errors and `SignalInterrupt` statuses are bubbled up to libbitcoinkernel callers.

I'd very much encourage adding a `util::Expected` class which is a simple wrapper over `std::variant` like your patch, if you see potential uses for it. But if you think adding `util::Expected` class would be overkill, and instead would prefer to build this functionality into `util::Result` that would also be fine with me. I just think it might make it harder to disentangle the different use-cases for `util::Result` and `std::expected` later when we update to c++23 if we are using one class instead of two.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r1665942681

> [79a970d](https://github.com/bitcoin/bitcoin/commit/79a970d4f12458a175317c453e251db7846c3561): I am not really sure, what the point of `Update` here is. A simple `return Interrupted{};` compiles as well. I fail to see why it would be wrong, because the default constructed monostate should not be returned either way.
> 
> (Same for all code below)
> 
> My recommendation would be to keep the code as simple as possible and not add complicated features or complication that isn't required.

I don't see a big difference in complexity here, but I'll make the suggested change since two people asked for it. Stickies suggested the same change in https://github.com/bitcoin/bitcoin/pull/25665#issuecomment-2079388456 and posted a diff.

For background, a good reason to use the `result.Update()` pattern is when you have a function performing multiple operations, and you want to return warnings and errors from all operations, not just the last one.

When I wrote more code using the `Result` class in #25722 and #29700, I found that whenever you have functions returning `Result` and calling other functions returning `Result`, a good way to write them was to declare a single `Result` variable at the top, and accumulate error and warning messages in it, and return it after updating it with a final success or failure value. Other patterns could work too, but this pattern was easy to follow, and worked in all cases including more complicated ones with loops, and nested if statements, and #ifdef code. In #25722 and #29700, for simpler `Result` functions that don't call other `Result` functions, I did use simpler pattern of just returning success values or `util::Error` directly, as you suggest.

In `CompleteChainstateInitialization` right now, either pattern works fine, because it does not currently call any other functions returning `Result`. I just thought `result.Update` pattern was appropriate because it is a pretty long function, and later in #29700 `LoadBlockIndex` and `MaybeRebalanceCaches` will be updated to return `Result` values, so using `result.Update` now could avoid code churn later.